### DISCUSSION TITLE
Stateful Fuzz(Open), what is the need of creating new instance on every run?

### DISCUSSION THREAD

**Discussion by sh0lt0** [Cyfrin/security-and-auditing-full-course-s23#151] [Q&A]
Stateful fuzz runs based on the previous state, why is a new instance created on every run 
` testCon Contract;
function setUp(){
testCon = new Contract();}`
I tried commenting the creation of new instance but the test showed no test to fuzz.

**Comment by EngrPips** [comment]
Hello @sh0lt0 , so i had to go re-run a stateful fuzzing test that i had initially written and read the stack trace of all the transaction that was executed during the stateful fuzz and i found out that the contract was only set up once for me and the remaining calls was between the handler and the invariant test i wrote. i am not sure why your own stateful fuzz is setting up new contract on every run. thats unusual maybe you are not doing something right.

**Comment by sh0lt0** [comment]
Hey, I meant stateful fuzz without handler, I am sharing the snippet of my test
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/22929138/f8a21e4a-f4da-49a0-966e-1fe65644c244)
 I have deliberately commented out the line where new instance is created inside setUp() but the on running the test it gives failed test `- no contract to fuzz` on the other hand, when I remove the comment, it runs fine.
 So, here my question arises what is the need of this line `statefullFuzzCont = new StatefulFuzzCatches();`

Apart from my initial question, I am curious how can I read my stack trace where you found out the calls and instance creation?

**Reply by EngrPips** [reply]
You didn't import `StdInvariant`, why is that ? . below is what my invariant test suite look like 

<img width="1440" alt="Screenshot 2024-04-05 at 2 53 24 PM" src="https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/95760918/4c05d807-86b5-4bc6-972e-615a1c5815ad">

You need to import `StdInvariant` to do stateful fuzzing.

**Reply by EngrPips** [reply]
The `security and auditing` course have a section where Patrick teaches how to write a proper invariant test suite. If you will follow this link `https://updraft.cyfrin.io/courses/security/tswap/stateful-and-stateless-fuzzing?lesson_format=video` it will take you the course on cyfrin updraft where Patrick teaches about `stateless fuzzing`, `open stateful fuzzing` and `controlled stateful fuzzing`

**Comment by sh0lt0** [comment]
Hey, thanks for your help, I think I need to rewatch the modules to get more clarity.
Could you please let me know how did you look at the stack trace please?

**Reply by EngrPips** [reply]
to look at the stack trace you can just put the `-vvvv` flag at the of your command to run your test

**Comment by sh0lt0** [comment]
I always run the command with max verbosity but I never looked it from a different perspective, thanks for the enlightening.

**Reply by EngrPips** [reply]
Anytime friend
